### DISCUSSION TITLE
zkSync Lite - paying for a transfer with a different fee token

### DISCUSSION THREAD

**Discussion by jtolio** [zkSync-Community-Hub/zksync-developers#618] [SDKs] [answered]
Hello!

For the last three years, every month, we (Storj Labs) have been paying our storage node operators over zkSync Lite with STORJ token. We also support vanilla Ethereum of course, and as of last year started support for Era as well.

This month, STORJ token is no longer listed as available for fee payment on https://zkscan.io/explorer/tokens/, which I understand to be a problem of trading volume. This is a big bummer, but not insurmountable. We are alright with paying fees for zkSync Lite in Ethereum or even USDC these days.

But I can't figure out how to do it! I can't find a zkSync Lite API that actually allows for specifying a different fee token for a transfer.

Here is specific code, found here: https://github.com/matter-labs/zksync/blob/master/sdk/zksync.js/src/utils.ts#L779

```
export function serializeTransfer(transfer: Transfer): Uint8Array {
    const type = new Uint8Array([255 - 5]);
    const version = new Uint8Array([CURRENT_TX_VERSION]);
    const accountId = serializeAccountId(transfer.accountId);
    const from = serializeAddress(transfer.from);
    const to = serializeAddress(transfer.to);
    const token = serializeTokenId(transfer.token);
    const amount = serializeAmountPacked(transfer.amount);
    const fee = serializeFeePacked(transfer.fee);
    const nonce = serializeNonce(transfer.nonce);
    const validFrom = serializeTimestamp(transfer.validFrom);
    const validUntil = serializeTimestamp(transfer.validUntil);
    return ethers.utils.concat([type, version, accountId, from, to, token, amount, fee, nonce, validFrom, validUntil]);
}
```

You'll notice that the token type is only included in the serialization once, and I don't know if there's a different serialization version that supports separated tokens for fee payment vs token transfer.

This is a relatively urgent request - we will need to fall back to layer 1 transfers for these payments if we can't resolve this in the next few days (layer 1 transfers have some additional restrictions due to higher transfer fees, so this isn't great).

Does anyone know how to either pay for a transfer with a different fee token via an API, or alternatively, know of an easy way to get STORJ supported for fee payment again?

Thanks!

**Comment by alhenaai** [comment]
Hello!

Thank you for reaching out with your query regarding the use of a different fee token for transactions on zkSync Lite. Based on the information you've provided and the current capabilities of zkSync Lite, here's what you need to know:

### Paying Fees with Different Tokens

In zkSync Lite, the ability to pay transaction fees using tokens other than the one being transferred (such as paying fees in ETH or USDC while transferring STORJ) is supported through the concept of "gasless meta-transactions." This means users can pay transaction fees in the tokens being transferred without needing to own ETH or any other specific tokens. However, this typically requires the token used for paying fees to have sufficient liquidity and be listed on platforms like Uniswap, and its price available on CoinGecko.

### Current Issue with STORJ

As you mentioned, STORJ is no longer listed as available for fee payment on zkScan, likely due to trading volume issues. This directly affects the ability to use STORJ for fee payments.

### Possible Solutions

1. **Using a Different Fee Token**: Currently, the zkSync Lite API and the standard transaction serialization do not explicitly support specifying a different token for the fee payment separate from the transfer token. The serialization function you've shared indeed only includes the token type once, which is used for both the transfer and the fee.

2. **Re-listing STORJ for Fee Payment**: To get STORJ supported again for fee payment, it would be necessary to ensure it meets the liquidity requirements on platforms like Uniswap and has a stable price listing on CoinGecko. This might involve engaging with the community or liquidity providers to boost the trading volume and visibility of STORJ.

3. **Fallback to Layer 1**: If re-enabling STORJ for fee payments on zkSync Lite is not feasible in the short term, and no API adjustments can be made promptly, using Layer 1 for transfers might be the only alternative, despite the higher fees and restrictions.

### Recommendation

Given the urgency of your situation, I recommend preparing for a potential temporary shift to Layer 1 transfers while simultaneously working on boosting STORJ's liquidity and trading volume to meet the requirements for fee payment on zkSync Lite. Additionally, reaching out directly to the zkSync support or developer community might provide more tailored advice or updates on potential upcoming features that could address your needs.

For any further assistance or to discuss this issue with the zkSync team, you might want to join the [zkSync developer community on Discord](https://join.zksync.dev/) where you can ask more detailed technical questions.

I hope this information helps, and you find a viable solution soon!

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by jtolio** [comment]
Whoa, an AI.

The AI mostly said back to me what I said, though I would specifically like to avoid falling back to layer 1, though it is the only option I know of for forward progress (we are not going to artificially change trading volumes). But before we do that, I hope that this message serves as me reaching out to the zkSync support and developer community directly. I would love a human's thoughts on this issue.

**Reply by bxpana** [reply]
Real human here. Will see what resources we have for this. Would it be better if Storj was just kept as the payment fee?

**Reply by jtolio** [reply]
Certainly that's easiest for us.

**Reply by JackHamer09** [reply, marked as answer]
**Regarding transfer and paying fee in a different token:**

This basically means making 2 transfers in a batch, where one of the transaction is your main transfer with 0 fee, and the other one is 0 amount transfer to yourself with fee token of your choice.

```ts
import { Wallet, getDefaultRestProvider, submitSignedTransactionsBatch, closestPackableTransactionFee } from "zksync";
import { BatchBuilder } from "zksync/build/batch-builder";

const provider = getDefaultRestProvider(network);
const wallet = Wallet.fromEthSigner(ethWallet, syncProvider);

// Transfer data
const transfer = {
  to: "0x...",
  token: "Storj",
  amount: "10000"
}
const feeTokenSymbol = "ETH";

// Getting batch transaction fee
let totalFee = await provider.getTransactionsBatchFee(["Transfer", "Transfer"], [transfer.address, wallet.address()], feeTokenSymbol);
totalFee = closestPackableTransactionFee(totalFee);

// Making transfer
const nonce = await wallet.getNonce("committed");
const batchBuilder = wallet.batchBuilder(nonce);

batchBuilder.addTransfer({
  to: transfer.to,
  token: transfer.symbol,
  amount: transfer.amount,
  fee: "0",
});
batchBuilder.addTransfer({
  to: wallet.address(),
  token: feeTokenSymbol,
  amount: "0",
  fee: totalFee,
});

const batchTransactionData = await batchBuilder.build();

const transactions = await submitSignedTransactionsBatch(
        syncWallet.provider,
        batchTransactionData.txs,
        batchTransactionData.signature ? [batchTransactionData.signature] : undefined,
);

// Awaiting for receipts (optional)
const receipts = await Promise.all(transactions.map((tx) => tx.awaitReceipt());
```

**Reply by jtolio** [reply]
We will need to port this code to Go to use it (https://github.com/storj/crypto-batch-payment/blob/main/pkg/zksync/zksync.go), so @bxpana, I am fairly interested in what your question was implying if that's on the table.

@JackHamer09 thank you for this though, I now know what to do lol

**Reply by bxpana** [reply]
Checking to see if we can get more info on why the token is no longer available for the fee token so we can maybe get it back as a fee token. Will keep you posted.

**Reply by jtolio** [reply]
Thank you so much!

**Reply by jtolio** [reply]
hey @bxpana - our clock is ticking down. considering the development effort above (and our intent to deprecate zksync lite support anyway), i think we're going to bail on zksync lite unless you have good news to report today. sorry for the pester but any update? thanks so much!

**Reply by bxpana** [reply]
@jtolio unfortunately no news to report today. You're not a pester at all. If I hear anything I'll let you know ASAP, but might not have an update before you deprecate Lite.

**Reply by jtolio** [reply]
Okay, thank you
